packages/datasets-fs/src/commands/generate.ts

```diff
@@ -0,0 +1,169 @@
+export type Args = z.infer<ArgsSchemaType>
+
+export abstract class Generate<T extends ArgsSchemaType> extends OutputCliCommand<T> {
+    public override getOptions() {
+        return lodash().merge(super.getOptions(), {
+            output: {
+                description: 'Where to write the captchas JSON file',
+            },
+            labelled: {
+                string: true,
+                demand: true,
+                description: 'Path to JSON file containing labelled data',
+            },
+            unlabelled: {
+                string: true,
+                demand: true,
+                description: 'Path to JSON file containing unlabelled data',
+            },
+            seed: {
+                number: true,
+                demand: true,
+                description: 'Seed for random number generator',
+            },
+            size: {
+                number: true,
+                description: 'Number of images in each captcha',
+            },
+            labels: {
+                string: true,
+                description:
+                    'Path to JSON file containing labels which unlabelled data will be assigned to. If not given, labels will be deduced from the labelled data.',
+            },
+            allowDuplicates: {
+                boolean: true,
+                description: 'If true, allow duplicates in the data (labelled and unlabelled)',
+            },
+            allowDuplicatesLabelled: {
+                boolean: true,
+                description: 'If true, allow duplicates in the labelled data',
+            },
+            allowDuplicatesUnlabelled: {
+                boolean: true,
+                description: 'If true, allow duplicates in the unlabelled data',
+            },
+        })
+    }
+
+    public override async _check(args: Args) {
+        // if specified, check files exist
+        const labelledMapFile: string | undefined = args.labelled
+        if (labelledMapFile && !fs.existsSync(labelledMapFile)) {
+            throw new ProsopoEnvError(
+                new Error(`labelled map file does not exist: ${labelledMapFile}`),
+                'FS.FILE_NOT_FOUND'
+            )
+        }
+        const unlabelledMapFile: string | undefined = args.unlabelled
+        if (unlabelledMapFile && !fs.existsSync(unlabelledMapFile)) {
+            throw new ProsopoEnvError(
+                new Error(`unlabelled map file does not exist: ${unlabelledMapFile}`),
+                'FS.FILE_NOT_FOUND'
+            )
+        }
+        this.labelledMapFile = labelledMapFile || ''
+        this.unlabelledMapFile = unlabelledMapFile || ''
+    }
+
+    labelled: LabelledItem[] = []
+    unlabelled: Item[] = []
+    labels: string[] = []
+    labelledMapFile = ''
+    unlabelledMapFile = ''
+    labelToImages: { [label: string]: Item[] } = {}
+    targets: string[] = []
+    saltRounds = 10
+
+    public override async _run(args: Args) {
```

Comment:
Pretty lengthy in places - could we break some parts of _run down into helper funcs like done with checkduplicated already on line 115?

Reply:
done